mdsobjects/cpp/mdstreeobjects.cpp

```diff
@@ -1661,7 +1681,307 @@ StringArray *TreeNodeArray::getUsage()
+      case NciK_IS_MEMBER:		return "\'IS_MEMBER\'";
+      case NciCONGLOMERATE_NIDS:	return "\'CONGLOMERATE_NIDS\'";
+      case NciORIGINAL_PART_NAME:	return "\'ORIGINAL_PART_NAME\'";
+      case NciNUMBER_OF_MEMBERS:	return "\'NUMBER_OF_MEMBERS\'";
+      case NciNUMBER_OF_CHILDREN :	return "\'NUMBER_OF_CHILDREN\'";
+      case NciMEMBER_NIDS:		return "\'MEMBER_NIDS\'";
+      case NciCHILDREN_NIDS:		return "\'CHILDREN_NIDS\'";
+      case NciFULLPATH:			return "\'FULLPATH\'";
+      case NciMINPATH:			return "\'MINPATH\'";
+      case NciUSAGE:			return "\'USAGE\'";
+      default:				throw MdsException("Unexpected NCI item in TreeNodeThinClient");
+    }
+}
+
+TreeNodeThinClient::TreeNodeThinClient(int nid, Connection *connection, Data *units, Data *error, Data *help, Data *validation)
+{
+    this->tree = 0;
+    this->nid = nid;
+    this->connection = connection;
+    clazz = CLASS_S;
+    dtype = DTYPE_NID;
+    setAccessory(units, error, help, validation);
+}
 
+std::string  TreeNodeThinClient::getNciString(int itm)
+{
+    char expr[64];
+    sprintf(expr, "GETNCI(%d,%s)", nid, convertNciItm(itm));
+    AutoData<Data> retStringData(connection->get(expr));
+    if(!retStringData.get())
+	throw MdsException("Error in Remote evaluation of getnci");
+    
+    AutoString as(retStringData->getString());
+    return as.string;
+}
+
+char TreeNodeThinClient::getNciChar(int itm)
+{
+    char expr[64];
+    sprintf(expr, "GETNCI(%d,%s)", nid, convertNciItm(itm));
+    AutoData<Data> retData(connection->get(expr));
+     if(!retData)
+	throw MdsException("Error in Remote evaluation of getnci");
+     char retChar = retData->getByte();
+     return retChar;
+}
+   
+int TreeNodeThinClient::getNciInt(int itm)
+{
+    char expr[64];
+    sprintf(expr, "GETNCI(%d,%s)", nid, convertNciItm(itm));
+    AutoData<Data> retData(connection->get(expr));
+    if(!retData)
+	throw MdsException("Error in Remote evaluation of getnci");
+    return retData->getInt();
+}
+
+int64_t TreeNodeThinClient::getNciInt64(int itm)
+{
+    char expr[64];
+    sprintf(expr, "GETNCI(%d,%s)", nid, convertNciItm(itm));
+    AutoData<Data> retData(connection->get(expr));
+    if(!retData)
+	throw MdsException("Error in Remote evaluation of getnci");
+    return retData->getLong();
+}
+
+char *TreeNodeThinClient::getPath()
+{
+    char expr[64];
+    sprintf(expr, "GETNCI(%d,\'PATH\')", nid);
+    AutoData<Data> retData(connection->get(expr));
+    if(!retData)
+	throw MdsException("Error in Remote evaluation of getnci(path)");
+    return retData->getString();
+}
+    
+EXPORT Data *TreeNodeThinClient::getData()
+{
+    char expr[64];
+    sprintf(expr, "GETNCI(%d,\'RECORD\')", nid);
+    Data *retData = connection->get(expr);
+    if(!retData)
+	throw MdsException("Error in Remote evaluation of getnci(record)");
+    return retData;
+}
+  
+EXPORT void TreeNodeThinClient::putData(Data *data)
+{
+    const char *path = (const char *)getPath();
+    Data *args[1];
```

Comment:
Still missing AutoData on these arrays of args (and something on the path on the line above).  Since get and put can throw, this will leak.  If passing the array of args is problematic, then you can use aggregation instead of composition to separate ownership:

AutoData<Data> d[] = { data->data() };
Data * args[] = { d[0].get() };

Otherwise, you need lots of try/catches, deletes, the delete in the for loop, etc.  Use C++ here to simplify.